driver-only ECDSA: get testing parity in SSL

valeriosetti opened the pull request:
## Description

The goal of this PR is to get test parity in driver-only build with accelerated ECDSA.
~Depends on #6997 to be merged~
Resolves #6860.

## Gatekeeper checklist

- [x] **changelog** not required
- [x] **backport** not required
- [X] **tests** provided

mpg commented on tests/suites/test_suite_ssl.data:
```diff
@@ -440,19 +440,19 @@ depends_on:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_AES_C:
 
 Handshake, select ECDH-RSA-WITH-AES-256-CBC-SHA384, non-opaque
-depends_on:MBEDTLS_HAS_ALG_SHA_384_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_AES_C:MBEDTLS_RSA_C:MBEDTLS_ECDSA_C:MBEDTLS_ECDH_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED
+depends_on:MBEDTLS_HAS_ALG_SHA_384_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_AES_C:MBEDTLS_RSA_C:MBEDTLS_PK_CAN_ECDSA_SOME:MBEDTLS_ECDH_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_ECP_DP_SECP384R1_ENABLED:MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED
```
Ah, that works for now but I'm afraid that's not strict enough in principle, which might affect future work (that is, when we work on builds with only public or only private operations). For the handshake to complete, one party need to make a signature and the other party needs to verify it, and since we act as both parties in these tests, we need both, so that's `MBEDTLS_PK_CAN_ECDSA_SIGN:MBEDTLS_PK_CAN_ECDSA_VERIFY`. (That is, sign _and verify, not _or_.)

For now the difference is not visible in our tests because we make no attempt to support builds with only public or only private operations, but there's demand for that (at least for builds with only public operations) so hopefully we'll get there at some point. And when we do, it will help if we did our best to use the right dependencies here :)

valeriosetti replied:
Is it worth to create a companion macro like `MBEDTLS_PK_CAN_ECDSA_FULL` to signal that there are both signing *and*  verifying capabilities?

mpg replied:
No sure. I think just listing them both is OK in tests: it's not much longer and those dependency lines are not short in the first place anyway. OTOH in other places it might be more convenient to have a single macro, for example in `ssl-opt.sh` to save vertical space by having on `requires_config_enabled` line instead of two.

valeriosetti replied:
Ok, so I will keep this change for the [following PR](https://github.com/Mbed-TLS/mbedtls/pull/7082) if it is necessary ;)

mpg commented on tests/suites/test_suite_ssl.function:
```diff
@@ -6,6 +6,7 @@
 #include <ssl_tls13_keys.h>
 #include <ssl_tls13_invasive.h>
 #include "test/certs.h"
+#include "pk_wrap.h"
```
Same point as before.

mpg reviewed (changes requested):
Looking pretty good, just one point carried over from X.509, and some nitpicking about future-proofing dependency declarations.

mpg commented:
Also, when 6997 is merged and this is rebased, please run `outcome-analysis.sh` as usual (but preferably not before, ideally it should be run on the version we're merging, to minimize risk).

valeriosetti commented:
- rebased after recent changes in #6997
- fixed changes requested by @mpg 
- tested with `outcome-analysis.sh` and attached results
[outcome-analysis-results.txt](https://github.com/Mbed-TLS/mbedtls/files/10721437/outcome-analysis-results.txt)

mpg commented on tests/src/certs.c:
```diff
@@ -25,6 +25,8 @@
 
 #include "mbedtls/legacy_or_psa.h"
 
+#include "pk_wrap.c"
```
This should also be `mbedtls/pk.h`. (Note: I'm assuming `.c` was a typo!)

valeriosetti replied:
sorry, this was an "impressive" (not in a good sense) series of errors all at once :/

mpg commented on tests/suites/test_suite_ssl.data:
```diff
@@ -3234,15 +3234,15 @@ depends_on:MBEDTLS_SSL_SESSION_TICKETS:MBEDTLS_SSL_CLI_C:MBEDTLS_SSL_PROTO_TLS1_
 
 Session serialization, save-load: no ticket, cert
-depends_on:MBEDTLS_X509_USE_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_ECDSA_C:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_FS_IO:MBEDTLS_SSL_PROTO_TLS1_2
+depends_on:MBEDTLS_X509_USE_C:MBEDTLS_PEM_PARSE_C:MBEDTLS_PK_CAN_ECDSA_SOME:MBEDTLS_ECP_DP_SECP256R1_ENABLED:MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA:MBEDTLS_FS_IO:MBEDTLS_SSL_PROTO_TLS1_2
```
Note to self: this does not perform a hanshake, so it doesn't need to generate or verify a signature, only to parse the certificate, so `ECDSA_SOME` is indeed appropriate.

mpg reviewed (changes requested):
Just one include left to fix, then I'm happy if the CI agrees.

valeriosetti commented:
Rebased after #6997 was merged into `development`

mpg reviewed (approved):
LGTM

mpg commented:
The component that failed in internal CI were infrastructure issues and the same components passed in the open CI, so this can be ignored for the purposes of merging this.
